### PR TITLE
feat: add language globs to findInFiles in napi

### PR DESCRIPTION
**Pull request by ppppqp**
Addressing this [issue](https://github.com/ast-grep/ast-grep/issues/745).

## Summary by CodeRabbit

- **New Features**
  - Introduced a new file search functionality supporting language-specific queries.
  - Added support for Vue single file components in the project.

- **Enhancements**
  - Expanded the file search capability to include TypeScript and HTML through language globs.

- **Bug Fixes**
  - Implemented error handling for unsupported and unknown language searches.

- **Platform Support**
  - Added native bindings support for the 'riscv64' architecture with distinction between 'musl' and 'gnu' libc variants.

- **Testing**
  - Created new tests to ensure correct file finding behavior with language globs and error handling.

- **Documentation**
  - Updated the project's documentation to reflect the addition of `LanguageGlobs` in file search configuration.

### CODE REVIEW

File: crates/napi/Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ edition.workspace = true
 [dependencies]
 ast-grep-core.workspace = true
 ast-grep-config.workspace = true
+ast-grep-language.workspace = true
```
Comment (HerringtonDarkholme) [inline crates/napi/Cargo.toml:16]: ast-grep-langauge has a lot of unnecessary language dependencies. say, tree-sitter parsers. 
I think we can skip the dependency here by implementing(copying) logic to `FrontEndLanguage`

File: crates/napi/__test__/index.spec.ts [hunk 2]
```diff
@@ -204,6 +204,54 @@ test('tsx should not find ts file', async t => {
   t.assert(true)
 })
 
+test('find with language globs', async t => {
+  countedPromise
```
Comment (HerringtonDarkholme) [inline crates/napi/__test__/index.spec.ts:208]: Looks like this is a typo?

File: crates/napi/__test__/index.spec.ts [hunk 3]
```diff
@@ -204,6 +204,54 @@ test('tsx should not find ts file', async t => {
...
+    paths: ['./'],
+    matcher: {
+      rule: {kind: 'member_expression'},
+    },
+    languageGlobs: {
+      'tsx': ['*.ts'],
+    }
+  }, (err, n) => {
+    t.assert(true)
+    t.is(err, null)
```
Comment (HerringtonDarkholme) [inline crates/napi/__test__/index.spec.ts:219]: Do we need this assertion?

File: crates/napi/index.d.ts [hunk 4]
```diff
@@ -46,6 +46,8 @@ export interface FindConfig {
   paths: Array<string>
   /** a Rule object to find what nodes will match */
   matcher: NapiConfig
+  /** find file by language (extension) */
```
Comment (HerringtonDarkholme) [inline crates/napi/index.d.ts:49]: IMHO the comment here can be more descriptive.

File: crates/napi/src/doc.rs [hunk 5]
```diff
@@ -50,6 +52,20 @@ impl Language for FrontEndLanguage {
...
+impl TryFrom<SupportLang> for FrontEndLanguage {
+  type Error = String;
+  fn try_from(s: SupportLang) -> Result<Self, Self::Error> {
+    match s {
+      SupportLang::Html => Ok(FrontEndLanguage::Html),
+      SupportLang::JavaScript => Ok(FrontEndLanguage::JavaScript),
+      SupportLang::Tsx => Ok(FrontEndLanguage::Tsx),
+      SupportLang::Css => Ok(FrontEndLanguage::Css),
+      SupportLang::TypeScript => Ok(FrontEndLanguage::TypeScript),
+      _ => Err(s.to_string())
```
Comment (HerringtonDarkholme) [inline crates/napi/src/doc.rs:64]: Is the error message only containing the language name? We can make it more helpful I think.

File: crates/napi/src/lib.rs [hunk 6]
```diff
@@ -302,19 +312,40 @@ fn select_custom<'b>(
   builder.select(file_type)
 }
 
-fn find_files_with_lang(paths: Vec<String>, lang: &FrontEndLanguage) -> Result<WalkParallel> {
+fn find_files_with_lang(paths: Vec<String>, language_globs: Option<LanguageGlobs>, lang: &FrontEndLanguage) -> Result<WalkParallel> {
```
Comment (HerringtonDarkholme) [inline crates/napi/src/lib.rs:315]: Can we swap the argument order? Just for cosmetic reason.

File: crates/napi/src/lib.rs [hunk 7]
```diff
@@ -302,19 +312,40 @@ fn select_custom<'b>(
...
     return Err(anyhow!("paths cannot be empty.").into());
   }
 
   let mut types = TypesBuilder::new();
   let types = types.add_defaults();
+
+  let mut custom_file_type = vec![];
+  if let Some(lgs) = language_globs {
+    for (l, globs) in lgs {
+      let result = SupportLang::from_str(&l);
```
Comment (HerringtonDarkholme) [inline crates/napi/src/lib.rs:326]: Can we use `?` operator here? anyhow has the `.context` extension method.

File: crates/napi/src/lib.rs [hunk 8]
```diff
@@ -302,19 +312,40 @@ fn select_custom<'b>(
...
+        return Err(anyhow!("unrecognized language in language globs").into());
+      }
+      let result = FrontEndLanguage::try_from(result.unwrap());
+      match result {
+        Ok(value) => {
+          if value == *lang{
+            custom_file_type.extend(globs);
+          }
+        },
+        Err(e) => return Err(anyhow!(format!("{} is not supported in napi", e)).into()),
```
Comment (HerringtonDarkholme) [inline crates/napi/src/lib.rs:337]: ditto, prefer anyhow extension

File: crates/napi/src/lib.rs [hunk 9]
```diff
@@ -302,19 +312,40 @@ fn select_custom<'b>(
...
   let types = match lang {
-    FrontEndLanguage::TypeScript => select_custom(types, "myts", &["*.ts", "*.mts", "*.cts"]),
-    FrontEndLanguage::Tsx => select_custom(types, "mytsx", &["*.tsx", "*.mtsx", "*.ctsx"]),
-    FrontEndLanguage::Css => types.select("css"),
-    FrontEndLanguage::Html => types.select("html"),
-    FrontEndLanguage::JavaScript => types.select("js"),
+    FrontEndLanguage::TypeScript => select_custom(types, "myts", vec!["*.ts", "*.mts", "*.cts"], custom_file_type),
+    FrontEndLanguage::Tsx => select_custom(types, "mytsx", vec!["*.tsx", "*.mtsx", "*.ctsx"], custom_file_type),
+    FrontEndLanguage::Css => select_custom(types, "css", vec!["css"], custom_file_type),
+    FrontEndLanguage::Html => select_custom(types, "html",  vec!["html"], custom_file_type),
```
Comment (HerringtonDarkholme) [inline crates/napi/src/lib.rs:347]: Can we use the same configuration here?

https://github.com/ast-grep/ast-grep/blob/main/crates/language/src/lib.rs#L288-L313

**Review by HerringtonDarkholme — CHANGES_REQUESTED** [review]
Hi! Thanks for your contribution to ast-grep. 

I have left some comments and look forward to your reply!

**Comment by ppppqp** [comment]
Thank you for your comments! Looking into my implementation more closely I found some questions.
I wonder how will [parse_files](https://github.com/ast-grep/ast-grep/blob/e82f620f136fbd5aafe744232d25315a1848882d/crates/napi/src/lib.rs#L226) be used and is it okay to change its signature?

**Comment by HerringtonDarkholme** [comment]
`parseFiles` is used like this 


```ts
parseFiles(['./*.ts'], (err, file) => {
  file.find('pattern')
})
```

I don't know why it is not included in [napi doc](https://ast-grep.github.io/reference/api.html#napi).

But changing its signature will be a breaking change.

**Comment by HerringtonDarkholme** [comment]
@coderabbitai review

**Review by HerringtonDarkholme — COMMENTED** [review]
Looks nice, would you like to run cargo fmt? The linter/formatter check failed.

**Review by HerringtonDarkholme — APPROVED** [review]

**Comment by HerringtonDarkholme** [comment]
Thanks!

**Pull request merged**